fix: private validator cache pubKey and reduce ping amount

JoeKash opened the pull request:
## Description

Reduce network usage when using remote signer

- reduced the ping time to wait 1 second before each successful ping (I've tested it with up to 10 seconds interval and couldn't find any issues so this can be reduced even more)

- cached the pubKey and don't keep requesting it from the remote (pubKey was requested dozen of times on a single block)

- Added some logs so it will be easier to debug such issues

I only did this cahnge for the 0.33.x branch but I can backport it to 0.32.x so it can be used on gaia as well

Closes: #5550

marbar3778 commented:
Thank you for opening this PR. We had a discussion internally and would like to avoid caching the pubkey as some applications handle key rotation or are looking at doing so. Could you make the change regarding the ping on master then backporting the change to the needed release can begin.

JoeKash commented:
Thanks @marbar3778  I'll look into making the ping changes on master.

Reducing the pings should be enough to solve this issue, but looking at the logs I added there are many requests to the PubKey in every block. 
Is there any chain that currently support dynamic changing of the public key ? 
How does it dynamically changes when using a local sign key ?
This is probably too big of a change to this fix but if the pubkey ever need to change it might be better to have an explicit method that will purge the cache, another option is to cache it just for the current block. I assume it won't and shouldn't change midblock signing ?

marbar3778 commented:
> but looking at the logs I added there are many requests to the PubKey in every block.
> Is there any chain that currently support dynamic changing of the public key ?
> How does it dynamically changes when using a local sign key ?
> This is probably too big of a change to this fix but if the pubkey ever need to change it might be better to have an explicit method that will purge the cache, another option is to cache it just for the current block. I assume it won't and shouldn't change midblock signing ?

We currently cache the key in consensus on master. This reduces the amount of calls, but you are right we should make the key only be cached for one block. We do not know of a project that supports key rotation, but we don't know all the users of Tendermint.

JoeKash commented:
@marbar3778 I've cut another version without catching the key, but it doesn't seem like it is able to sync up.
It seem my first impression of the node asking the key dozen of times was off and it is actually a few hundred times between each block (See log attached) 

If this on master can't it backport ? Do you have a link to where this was discussed and decided against ?
If there is no option to change the pubKey locally so it's esentially cached won't it make sense it won't be changed when using a remote signer ?

If you are still against this, can we backport it with a config option disabling cache by default but allowing to enable it ?

```log
I[2020-11-03|04:50:37.822] Executed block                               module=state height=202798 validTxs=0 invalidTxs=0
I[2020-11-03|04:50:37.886] Committed state                              module=state height=202798 txs=0 appHash=26359A282A2929D49F50248C2AA21775B1443F41ED64B006BA22A781F39677B4
I[2020-11-03|04:50:37.890] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:37.890] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:37.951] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:37.951] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:37.951] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:38.012] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:38.012] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:38.012] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:38.073] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:38.073] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:38.073] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:38.134] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:38.134] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:38.134] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:38.195] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:38.195] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:38.195] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:38.255] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:38.255] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:38.255] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:38.316] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:38.316] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:38.316] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:38.361] SignerListener: PingLoop Active              module=privval
I[2020-11-03|04:50:38.483] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:38.483] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:38.483] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:38.544] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:38.544] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:38.544] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:38.605] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:38.605] SignerListener: PingLoop successes - sleeping module=privval
I[2020-11-03|04:50:38.605] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:38.665] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:38.665] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:38.665] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:38.726] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:38.726] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:38.726] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:38.787] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:38.787] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:38.787] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:38.848] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:38.848] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:38.848] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:38.909] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:38.909] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:38.909] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:38.969] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:38.969] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:38.969] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:39.030] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:39.030] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:39.030] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:39.091] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:39.091] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:39.091] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:39.160] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:39.160] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:39.160] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:39.222] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:39.222] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:39.222] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:39.283] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:39.283] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:39.283] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:39.344] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:39.344] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:39.344] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:39.405] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:39.405] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:39.405] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:39.466] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:39.466] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:39.466] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:39.527] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:39.527] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:39.527] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:39.587] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:39.587] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:39.587] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:39.605] SignerListener: PingLoop Active              module=privval
I[2020-11-03|04:50:39.648] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:39.648] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:39.648] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:39.709] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:39.709] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:39.709] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:39.770] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:39.770] SignerListener: PingLoop successes - sleeping module=privval
I[2020-11-03|04:50:39.770] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:39.831] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:39.831] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:39.831] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:39.892] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:39.892] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:39.892] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:39.952] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:39.952] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:39.952] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:40.013] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:40.013] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:40.013] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:40.074] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:40.074] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:40.074] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:40.135] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:40.135] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:40.135] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:40.196] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:40.196] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:40.196] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:40.257] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:40.257] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:40.257] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:40.318] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:40.318] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:40.318] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:40.378] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:40.378] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:40.378] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:40.439] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:40.439] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:40.439] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:40.500] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:40.500] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:40.500] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:40.561] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:40.561] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:40.561] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:40.622] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:40.622] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:40.622] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:40.685] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:40.685] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:40.685] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:40.745] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:40.745] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:40.745] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:40.770] SignerListener: PingLoop Active              module=privval
I[2020-11-03|04:50:40.806] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:40.806] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:40.806] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:40.867] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:40.867] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:40.867] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:40.928] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:40.928] SignerListener: PingLoop successes - sleeping module=privval
I[2020-11-03|04:50:40.928] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:40.989] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:40.989] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:40.989] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:41.050] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:41.050] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:41.050] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:41.111] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:41.111] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:41.111] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:41.172] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:41.172] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:41.172] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:41.232] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:41.232] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:41.232] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:41.293] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:41.293] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:41.293] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:41.354] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:41.354] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:41.354] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:41.414] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:41.414] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:41.414] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:41.475] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:41.475] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:41.475] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:41.537] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:41.537] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:41.537] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:41.597] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:41.597] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:41.597] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:41.659] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:41.659] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:41.659] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:41.719] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:41.719] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:41.719] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:41.780] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:41.780] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:41.780] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:41.841] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:41.841] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:41.841] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:41.902] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:41.902] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:41.902] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:41.928] SignerListener: PingLoop Active              module=privval
I[2020-11-03|04:50:42.063] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:42.063] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:42.063] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:42.123] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:42.123] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:42.123] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:42.184] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:42.184] SignerListener: PingLoop successes - sleeping module=privval
I[2020-11-03|04:50:42.184] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:42.245] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:42.245] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:42.245] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:42.306] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:42.306] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:42.306] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:42.367] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:42.367] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:42.367] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:42.428] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:42.428] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:42.428] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:42.489] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:42.489] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:42.489] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:42.549] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:42.549] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:42.549] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:42.610] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:42.610] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:42.610] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:42.671] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:42.671] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:42.671] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:42.731] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:42.731] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:42.731] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:42.792] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:42.792] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:42.792] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:42.853] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:42.853] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:42.853] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:42.914] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:42.914] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:42.914] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:42.974] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:42.974] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:42.974] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:43.035] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:43.035] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:43.035] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:43.096] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:43.096] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:43.096] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:43.157] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:43.157] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:43.157] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:43.184] SignerListener: PingLoop Active              module=privval
I[2020-11-03|04:50:43.218] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:43.218] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:43.218] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:43.278] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:43.278] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:43.278] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:43.339] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:43.339] SignerListener: PingLoop successes - sleeping module=privval
I[2020-11-03|04:50:43.339] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:43.400] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:43.400] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:43.400] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:43.461] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:43.461] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:43.461] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:43.522] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:43.522] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:43.522] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:43.582] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:43.582] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:43.582] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:43.643] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:43.643] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:43.643] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:43.704] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:43.704] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:43.704] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:43.765] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:43.765] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:43.765] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:43.826] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:43.826] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:43.826] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:43.886] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:43.886] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:43.886] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:43.947] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:43.947] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:43.947] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:44.008] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:44.008] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:44.008] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:44.069] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:44.069] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:44.069] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:44.129] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:44.129] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:44.129] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:44.190] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:44.190] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:44.190] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:44.251] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:44.251] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:44.251] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:44.312] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:44.312] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:44.312] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:44.339] SignerListener: PingLoop Active              module=privval
I[2020-11-03|04:50:44.473] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:44.473] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:44.473] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:44.533] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:44.533] SignerListener: PingLoop successes - sleeping module=privval
I[2020-11-03|04:50:44.533] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:44.594] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:44.744] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:44.744] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:44.805] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:44.807] SignerClient: GetPubKey                      module=privval
I[2020-11-03|04:50:44.807] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:44.868] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:44.868] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:44.929] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:45.536] SignerListener: PingLoop Active              module=privval
I[2020-11-03|04:50:45.536] SignerListener: SendRequest sent             module=privval
I[2020-11-03|04:50:45.598] SignerListener: SendRequest got response     module=privval
I[2020-11-03|04:50:45.598] SignerListener: PingLoop successes - sleeping module=privval
I[2020-11-03|04:50:46.465] Executed block                               module=state height=202799 validTxs=0 invalidTxs=0
I[2020-11-03|04:50:46.563] Committed state                              module=state height=202799 txs=0 appHash=2C960CFE2D7F6B574C84D5CAA902ED6B9F2705D71907E2B6EFA2D01973786C76
```

marbar3778 commented:
> It seem my first impression of the node asking the key dozen of times was off and it is actually a few hundred times between each block (See log attached)

This is odd, it should not be the case. It looks like the privval module is not handling requests correctly. 

> If this on master can't it backport ? Do you have a link to where this was discussed and decided against ?

We can look at backporting the change on master. But if you would like to backport the ping delay you will have to make a PR to master. This PR is set to the incorrect branch.

> If there is no option to change the pubKey locally so it's esentially cached won't it make sense it won't be changed when using a remote signer ?

I believe you can change the pubkey at anytime. By either replacing the priv_validator_key.json or the remote signer to feeding a different key.

JoeKash commented:
I rebased the changes to master 

> I believe you can change the pubkey at anytime. By either replacing the priv_validator_key.json or the remote signer to feeding a different key.

I don't think changing the priv_validator_key.json while the node is running will have any effect. you'll need to restart the node, which is identical to the behavior with cache.

I guess with remote signer you can do what ever your remote support, but I would imagine you want the validator nodes to be able to operate without remote as well, and the way it currently works you can't change the pubkey while the node is running.

So I would expect that if changing the key is a requirement, that will be also added to the local key.

Also I might be missing something, but I don't see how caching is preventing key rotation ?
Looking at this proposal https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-016-validator-consensus-key-rotation.md
It seem like a message will need to be sent to change the signing key. But this is all done outside of tendermint. 
You can't just expect the key to change after X blocked being signed so tmkms / tendermint don't know when it need to be changed.

So you need either a way to ask tendermint to generate a new key or have a way to tell tendermint the key the new key (but that's not really safe as keys should not be handled by the sdk and the sdk should probably not talk to the remote signer)
And you need to allow for some transitional period where the old key is still used or if the change fails.

So if I need to think of a safe workflow here -
1. sdk sends a request for a new signing key from tendermint
2. tendermint generates a new key (or pass the request to the remote signer to get a key)
3. new pubKey is sent to the sdk and is added to a `MsgRotateConsPubKey` and broadcast
4. once message is added to a block and successful the sdk request tendermint to change the consensus key .
5. tendermint changes the consensus key (or pass the command to the remote signer) 

I don't know if all this logic exists on tendermint, but if it does it should be very easy to clear the pubKey cache on step 5

melekes commented:
It's possible to cache the key and clear the cache once Tendermint receives an `EndBlock` from the ABCI application with current node's pubkey removed from the set.

That said, why are you concerned with the amount of requests?

JoeKash commented:
> That said, why are you concerned with the amount of requests?

@melekes see https://github.com/tendermint/tendermint/issues/5550 
We can't use tmkms as all this requests are blocking and causing the node to lag behind the network.

> I don't think we do that anymore.

You can see from my log above that this is still the case for tendermint 0.33.x and 0.32.x so any network using them (which includes Cosmos) will suffer from this issue.

melekes commented:
@JoeKash I see. Have you tried to cache the pubkey and can you confirm that the issue goes away once you do that (i.e. reducing the number of requests from 3 to 2 per block)? sorry, I am catching up with this issue.

JoeKash commented:
Yes @melekes. this PR has two commits the first one fixes both ping and caches the pubKey. 
As @marbar3778  request I removed the caching for the pubKey (second commit).
Using this fix (with the pubKey cache) tendermint 0.33.x solves my syncing issue.
My implementation just caches the pubKey after first successful response and does not allow it to be changed without restarting the node.But we can easily add a method to clear the cache and once the implementation to allow for validator pubKey changing is added it can use it to clear the cache

melekes commented:
Do you understand why caching solved your issue? Is it connected to the number of requests? Or maybe to particular GetPubKey request?

JoeKash commented:
Each one of those requests is blocking the node from doing actual work.
At the beginig I though that just removing the spam ping requests will be enough to solve this issue.
But that only helped a bit (I can  get the node to sign about 10% of the blocks if I don't constantly ping tmkms)
This about it you have more than 100 requests going for pubKey (1 for each validator + a few extra) and each takes at best 70ms so you lost 7 seconds and your chance to sign a block just for asking the tmkms for a value you already have.

melekes commented:
Closing for https://github.com/tendermint/tendermint/pull/5638. Thank you @JoeKash for your contribution!
